aggregator/src/aggregator/collection_job_driver.rs

```diff
@@ -390,28 +396,22 @@ where
         // job URI can serve it up. Scrub the batch aggregations, as we are now done with them, too.
         let collection_job = Arc::new(
             collection_job.with_state(CollectionJobState::Finished {
-                report_count,
-                client_timestamp_interval,
+                report_count: leader_aggregate_share.report_count,
+                client_timestamp_interval: leader_aggregate_share.client_timestamp_interval,
                 encrypted_helper_aggregate_share: AggregateShare::get_decoded(http_response.body())
                     .map_err(Error::MessageDecode)?
                     .encrypted_aggregate_share()
                     .clone(),
-                leader_aggregate_share,
+                leader_aggregate_share: leader_aggregate_share.aggregate_share,
             }),
         );
-        let batch_aggregations = Arc::new(
-            batch_aggregations
-                .into_iter()
-                .map(BatchAggregation::scrubbed)
-                .collect::<Vec<_>>(),
-        );
 
         datastore
             .run_tx("step_collection_job_2", |tx| {
                 let vdaf = Arc::clone(&vdaf);
                 let lease = Arc::clone(&lease);
                 let collection_job = Arc::clone(&collection_job);
-                let batch_aggregations = Arc::clone(&batch_aggregations);
+                let batch_aggregations = all_bas_iter.clone();
```

Comment:
I think we should wrap this in an `Arc` before passing it into this transaction, to avoid additional expensive copies.

Reply:
Whoops, I take it back! The interior mutability of a mutex is a problem because the transaction may need a retry, and thus the second attempt would see an empty iterator.

One other idea I had was to scrub the batch aggregations without having any corresponding list of `BatchAggregation`s in memory, either constructing them on the fly from the same Cartesian product, or just writing a separate UPDATE statement to NULL out the appropriate columns in all applicable rows at once.

Reply:
I think we can make this work while avoiding the cost of copying the BA values around by using `std::borrow::Cow<BatchAggregation<_>>` for the real BAs, and reconstructing the empty ones on the fly. That way we can work with references to the real BAs but still have a single type for the `BatchAggregationsIterator` to yield. The other thing I had to do was re-create the `BatchAggregationsIterator` in either `step_collection_job` closure.